sklearn/utils/_array_api.py

```diff
@@ -167,19 +210,52 @@ def _isdtype_single(dtype, kind, *, xp):
         return dtype == kind
 
 
-def supported_float_dtypes(xp):
-    """Supported floating point types for the namespace
+@lru_cache
```

Comment:
Another option: can `array-api-compat` help make all devices hashable?

Reply:
We would first need to be 100% sure that device objects are required to be hashable in the Array API spec itself.

The current state of the spec does not seem to make any explicit statement in that regard: https://data-apis.org/array-api/latest/design_topics/device_support.html#syntax-for-device-assignment

Reply:
We can wrap the `device` object in a hashable object and then pass that as input of the function that is `@lru_cache` 'd. The hash can be derived from `xp.__name__` and `str(device)` for instance.

But didn't we test this branch with cupy ? does it fail at the moment ? if it doesn't fail it means we lack coverage ? because we thought of non-hashable device objects already for the `device` function (see https://github.com/scikit-learn/scikit-learn/pull/27904/files#diff-d0cf1401f90f9cff12a7dd9778bafb5d1a06137f4145be07c8530e2502af131fR263 )

Reply:
I just realised I don't know the original motivation for using `lru_cache` here. Maybe Franck(?) can give a short explanation. Just to make sure there isn't an alternative solution/it is indeed needed. Before we embark on finding a solution to the hashing problem.

I also had a sense of deja vu regarding hashing of device objects, but I couldn't find where it related to. Thanks for the link Franck. I found this because I was running the tests on a machine that has cupy installed. So I think we have enough test coverage, but we lack "regularly running all the tests with all the libraries" (the solution is CI).

I think my favourite option would be "avoid the need for `lru_cache`" or making our own caching mechanism that does not rely on hashable objects (I think the lack of constant time look up isn't a big problem because per run there will only be a couple of entries in the cache? To be checked). It seems all other options are some form of "hack", either because we add some shim in scikit-learn or in `array-api-compat`. The standard is, on purpose, very vague on what the `Device` objects should/shouldn't do (so I guess we have to be prepared for "anything goes").

Reply:
Maybe the overhead is negligible and we can do away without cache for now and profile from time to time check that this helper is not introducing too much overhead in real life usage.

Reply:
With `xp` being `torch`, I have:

```
In [11]: %timeit _supports_dtype(xp, "cuda", "float16")
36.4 µs ± 74.4 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
```

without `lru_cache`, and

```
In [15]: %timeit _supports_dtype(xp, "cuda", "float16")
81.3 ns ± 0.237 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
```

with `lru_cache`. I'll keep the lru_cache (after debugging and testing it) in the next commit for the time being but feel free to remove it if you find the added complexity is not worth the small addition in reliability.

Reply:
I'd vote for removing it for now. It gets us unblocked on this PR and we can always add it back later. Especially because the road forward with `lru_cache` isn't clear to me as (at least) Cupy's `Device` isn't hashable.